codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt

```diff
@@ -993,12 +1008,28 @@ private class ServerHttpBoundProtocolTraitImplGenerator(
                 // `<_>::from()` is necessary to convert the `&str` into:
                 //     * the Rust enum in case the `string` shape has the `enum` trait; or
                 //     * `String` in case it doesn't.
-                rustTemplate(
+                when (type.rustType()) {
+                    RustType.String ->
+                        rustTemplate(
+                            """
+                            let value = <#{T}>::from(#{PercentEncoding}::percent_decode_str(value).decode_utf8()?.as_ref());
+                            """,
+                            *codegenScope,
+                            "T" to type,
+                        )
+                    else -> // RustType.Opaque, the Enum
```

Comment:
Add a `check` assertion to ensure it's `RustType.Opaque`, or use `RustType.Opaque` in this arm and then `throw` an exception in the `else` arm. Always code defensively to avoid introducing future bugs :)